src/btreemap/iter.rs

```diff
@@ -140,21 +128,67 @@ where
                     },
                     node,
                 });
+            }
+        }
+    }
+
+    fn next_internal(&mut self, load_value: bool) -> Option<(K, Option<V>)> {
+        // If the cursors are empty. Iteration is complete.
+        let cursor_last = self.cursors.pop()?;
+        self.update_cursors_for_iteration(cursor_last.clone());
+        if let Cursor::Node {
+            node,
+            next: Index::Entry(entry_idx),
+        } = cursor_last
+        {
+            if entry_idx < node.entries_len() {
+                let key = node.key(entry_idx);
 
                 // If the key does not belong to the range, iteration stops.
-                if !self.range.contains(&key) {
+                if !self.range.contains(key) {
                     // Clear all cursors to avoid needless work in subsequent calls.
                     self.cursors = vec![];
                     return None;
                 }
-
-                Some((key, V::from_bytes(Cow::Owned(encoded_value))))
-            }
-            None => {
-                // The cursors are empty. Iteration is complete.
-                None
+                return Some((
+                    key.clone(),
+                    if load_value {
+                        let encoded_value = node.value(entry_idx, self.map.memory()).to_vec();
+                        Some(V::from_bytes(Cow::Owned(encoded_value)))
+                    } else {
+                        None
+                    },
+                ));
             }
         }
+
+        self.next_internal(load_value)
+    }
+}
+
+impl<K, V, M> Iterator for Iter<'_, K, V, M>
+where
+    K: Storable + Ord + Clone,
+    V: Storable,
+    M: Memory,
+{
+    type Item = (K, V);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // It is safe to unwrap value because when the key exists the
+        // value will exist as well, and it will be loaded.
+        self.next_internal(true).map(|(k, v)| (k, v.unwrap()))
```

Comment:
This is a step in the right direction, but I'm not a big fan of having unwraps :) There's a way to avoid unwraps as well as cloning. I'll let you think about it - ping me if you get stuck.

Reply:
Cloning is removed, but for the unwraps I am not sure in which way you want me to continue, can you please tell me?

Reply:
Sure, here's what I'd do:

Instead of `next_internal`, I'd define a method called `next_map`. This is a method that iterates through the tree, and the caller can provide a `map` closure that takes a node and entry point idx and maps it to whatever it wants to return. Here's the signature:

```
fn next_map<T, F: Fn(&Node<K>, usize) -> T>(&mut self, map: F) -> Option<T> {
```

Then, `next` can be implemented like this:

```
impl<K, V, M> Iterator for Iter<'_, K, V, M>
where
    K: Storable + Ord + Clone,
    V: Storable,
    M: Memory,
{
    type Item = (K, V);

    fn next(&mut self) -> Option<Self::Item> {
        self.next_map(|node, entry_idx| {
            let (key, encoded_value) = node.entry(entry_idx, self.map.memory());
            (key, V::from_bytes(Cow::Owned(encoded_value)))
        })
    }
}
```

The count can be implemented similarly, but the map function will simply do nothing.